pkgs/by-name/pa/papers/package.nix

```diff
@@ -0,0 +1,161 @@
+{ lib
+, stdenv
+, fetchFromGitLab
+, meson
+, ninja
+, pkg-config
+, appstream
+, desktop-file-utils
+, gtk4
+, glib
+, pango
+, atk
+, gdk-pixbuf
+, shared-mime-info
+, itstool
+, gnome
+, poppler
+, ghostscriptX
+, djvulibre
+, libspectre
+, libarchive
+, libsecret
+, wrapGAppsHook
+, librsvg
+, gobject-introspection
+, yelp-tools
+, gspell
+, gsettings-desktop-schemas
+, dbus
+, gst_all_1
+, gi-docgen
+, supportMultimedia ? true # PDF multimedia
+, libgxps
+, supportXPS ? true # Open XML Paper Specification via libgxps
+, withLibsecret ? true
+, libadwaita
+, exempi
+, cargo
+, rustPlatform
+}:
+
+stdenv.mkDerivation rec {
+  pname = "papers";
+  version = "unstable-2024-3-27";
+
+  outputs = [ "out" "dev" "devdoc" ];
+
+  src = fetchFromGitLab {
+    domain = "gitlab.gnome.org";
+    owner = "GNOME/Incubator";
+    repo = "papers";
+    rev = "4374535f4f5e5cea613b2df7b3dc99e97da27d99";
+    hash = "sha256-wjLRGENJ+TjXV3JPn/lcqv3DonAsJrC0OiLs1DoNHkc=";
+  };
+
+  cargoDeps = rustPlatform.importCargoLock {
+    lockFile = "${src}/shell-rs/Cargo.lock";
```

Comment:
You cannot import stuff from a derivation, IFD is not allowed in Nixpkgs.

Reply:
I don't understand what you mean? How can I get the rust dependencies if I don't use "importCargoLock"?

Reply:
You would need to copy the `Cargo.toml` into the Nixpkgs repository, or use `rustPlatform.fetchCargoTarball` instead, which does stuff at the build time rather than at evaluation time, at the cost of the dependencies being stuffed into a single fixed-output derivation.

Reply:
I have used "importCargoLock" because some dependencies were git repositories inside Cargo.lock and I was given an error message stating it is unsupported by fetchCargoTarball. What is the correct way to do so ? Ignore the Cargo.lock ?

Reply:
@jtojnar 
I really don't understand how to package using "fetchCargoTarball" rather than "ImportCargoLock". Everytime I try, I'm stuck with this error message:

```
error: builder for '/nix/store/w3s1ghln36l8czzsqk852ib1cq0qbywb-papers-46.0-unstable-2024-03-27-vendor.tar.gz.drv' failed with exit code 1;
       last 10 log lines:
       > source = "git+https://gitlab.gnome.org/World/Rust/libadwaita-rs.git?branch=master#9f9b3e8026f3cf7d7f1095fe5d0599f4fa78f625"
       > source = "git+https://gitlab.gnome.org/World/Rust/libadwaita-rs.git?branch=master#9f9b3e8026f3cf7d7f1095fe5d0599f4fa78f625"
       > source = "git+https://github.com/gtk-rs/gtk-rs-core?branch=master#8e5054e4daea31be1fe2843d4003a00ccc8b6451"
       > source = "git+https://github.com/gtk-rs/gtk-rs-core?branch=master#8e5054e4daea31be1fe2843d4003a00ccc8b6451"
       >
       > ERROR: The Cargo.lock contains git dependencies
       >
       > This is currently not supported in the fixed-output derivation fetcher.
       > Use cargoLock.lockFile / importCargoLock instead.
       >
       For full logs, run 'nix log /nix/store/w3s1ghln36l8czzsqk852ib1cq0qbywb-papers-46.0-unstable-2024-03-27-vendor.tar.gz.drv'.
error: 1 dependencies of derivation '/nix/store/si8gzmz3n8mn0yjzl7r6c50s6l2h2pcv-papers-46.0-unstable-2024-03-27.drv' failed to build
```

The problem is, the versions used of these dependencies just don't exist inside the cargo registry. So I'm forced to use the git version.

Is there a solution I don't know to package them properly ?

Reply:
You will need to use `importCargoLock` with the lockfile copied to the repo, see other uses of the function in Nixpkgs.

Reply:
I've added the Cargo.lock inside the nixpkgs repository with today's commit. Is it better now ?